### PR TITLE
Design for Key Derivation

### PR DESCRIPTION
**Pull request by otherview**
### Why is this change needed?

- Design for https://github.com/obscuronet/obscuro-internal/issues/1193

### What changes were made as part of this PR:

- Provide a high level list of the changes made
- List key areas for the reviewer 

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: design/key_derivation.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+* Each rollup is encrypted with a unique key
+  * Each unique key is derived from the network key
+  * Each unique key is not able to disclose other unique keys
+  * Each unique key is not able to disclose the network key
+  * Each unique key is deterministic give the rollup height
+* Each enclave is able to determine the unique key given the enclave secret and rollup height
+
+## Design
+
+Use HKDF standardized in https://www.rfc-editor.org/rfc/rfc5869.
```
Comment (tudor-malene) [inline design/key_derivation.md:23]: I'd suggest expanding this section a bit. 
First a bit of context.
Followed by the options.
Each with pros, cons.

File: design/key_derivation.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+This key is derived from the network Key.
+
+## Requirements
+
+* Each rollup is encrypted with a unique key
+  * Each unique key is derived from the network key
+  * Each unique key is not able to disclose other unique keys
+  * Each unique key is not able to disclose the network key
+  * Each unique key is deterministic give the rollup height
+* Each enclave is able to determine the unique key given the enclave secret and rollup height
```
Comment (tudor-malene) [inline design/key_derivation.md:19]: the requirement is to encrypt with symmetric keys.
Also need a bit of analysis of the existing symmetric algorithms.

File: design/key_derivation.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Key Derivation design
+
+## Scope
+
+Obscuro blockchain is seeded with a network Key.
```
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:5]: Good point - Rephared it.

File: design/key_derivation.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+This network key is shared between enclaves upon successful attestation.
```
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:6]: Removed master seed concept.

File: design/key_derivation.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+Transactions are encrypted in the rollups that make the Obscuro blockchain.
+Each rollup uses a key to encrypt the Transactions.
```
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:9]: I was scoping out the revelation period. But will add.

File: design/key_derivation.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+This key is derived from the network Key.
+
+## Requirements
+
+* Each rollup is encrypted with a unique key
```
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:14]: Good point 👍

File: design/key_derivation.md (hunk 1 above)
Comment (otherview) [reply design/key_derivation.md:23]: Can do 👍

File: design/key_derivation.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  * Each unique key is derived from the network key
+  * Each unique key is not able to disclose other unique keys
+  * Each unique key is not able to disclose the network key
+  * Each unique key is deterministic give the rollup height
```
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:18]: I've added a sentence stating the de-scoping of the revelation period to this document. I think once we have nailed the key derivation down we can expand or compose this document with the revelation period constraints.

File: design/key_derivation.md (hunk 2 above)
Comment (otherview) [reply design/key_derivation.md:22]: Added more info 👍 

I think for the first iteration AES is the right tool. It's the industry standard.
I did do a more in-depth on what type of cipher modes we should use. I think that's pretty interesting.

File: design/key_derivation.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
+# Key Derivation design
+
+## Scope
+
+The Obscuro blockchain has a network key that allows read/write to the transaction in the block chain.
+This network key is shared between enclaves upon successful attestation.
+
+Transactions are encrypted in the rollups that make the Obscuro blockchain.
+Each rollup uses a key to encrypt the Transactions.
+Each key is derived from the initial network Key.
```
Comment (tudor-malene) [inline design/key_derivation.md:10]: it's derived from the master seed which is not a key

Comment (otherview) [reply design/key_derivation.md:10]: gotcha

File: design/key_derivation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md:6]: I though it was the network shared key ?

[the remark this answers is not in the document]
Comment (otherview) [reply design/key_derivation.md]: Changed to master seed.

**Pull request closed without merging**